modules/widgets/instagram.php

```diff
@@ -0,0 +1,497 @@
+
+		return $is_legacy_token;
+	}
+
+	private function get_token_status( $token_id ) {
+		if ( empty( $token_id ) ) {
+			return [ 'valid' => false ];
+		}
+		if ( defined( 'IS_WPCOM' ) && IS_WPCOM ) {
+			$token = Keyring::init()->get_token_store()->get_token( array( 'type' => 'access', 'id' => $token_id ) );
+
+			return [
+				'valid' => ! empty( $token ),
+				'legacy' => $token && 'instagram' === $token->name,
+			];
+		}
+
+		$site = Jetpack_Options::get_option( 'id' );
+		$path = sprintf( '/sites/%s/instagram/%s/check-token', $site, $token_id );
+		$result = $this->wpcom_json_api_request_as_blog( $path, 2, array( 'headers' => array( 'content-type' => 'application/json' ) ), null, 'wpcom' );
+		$response_code = wp_remote_retrieve_response_code( $result );
+		if ( 200 !== $response_code ) {
+			return [
+				//We assume the token is valid if the response_code is anything but the invalid
+				//token codes we send back. This is to make sure it's not reset, if the API is down
+				//or something
+				'valid' => ! ( 403 == $response_code || 401 === $response_code ),
+				'legacy' => 'ERROR',
+			];
+		}
+		$status = json_decode( $result['body'], true );
+		return $status;
+	}
+
+	/**
+	 * Validates the widget instance's token ID and then uses it to fetch images from Instagram.
+	 * It then caches the result which it will use on subsequent pageviews.
+	 * Keyring is not loaded nor is a remote request is not made in the event of a cache hit.
+	 *
+	 * @param array $instance A widget $instance, as passed to a widget's widget() method.
+	 * @return string|array A string on error, an array of images on success.
+	 */
+	public function get_data( $instance ) {
+		if ( empty( $instance['token_id'] ) ) {
+			return 'ERROR';
+		}
+
+		$cache_time = MINUTE_IN_SECONDS;
+		$transient_key = implode( '|', array( 'instagram-widget', $instance['token_id'], $instance['count'] ) );
+		$cached_images = get_transient( $transient_key );
+		if ( $cached_images ) {
+			return $cached_images;
+		}
+
+		$site = Jetpack_Options::get_option( 'id' );
+		$path = sprintf( '/sites/%s/instagram/%s?count=%s', $site, $instance['token_id'], $instance['count'] );
+		$result = $this->wpcom_json_api_request_as_blog( $path, 2, array( 'headers' => array( 'content-type' => 'application/json' ) ), null, 'wpcom' );
+
+		$response_code = wp_remote_retrieve_response_code( $result );
+		if ( 200 !== $response_code ) {
+			set_transient( $transient_key, 'ERROR', $cache_time );
+			return 'ERROR';
+		}
+
+		$data = json_decode( wp_remote_retrieve_body( $result ), true );
+		if ( ! isset( $data['images'] ) || ! is_array( $data['images'] ) ) {
+			set_transient( $transient_key, 'ERROR', $cache_time );
+			return 'ERROR';
+		}
+
+		$cache_time = 20 * MINUTE_IN_SECONDS;
+		set_transient( $transient_key, $data, $cache_time );
+		return $data;
+	}
+
+	private function wpcom_json_api_request_as_blog( $path, $version = 1, $args = array(), $body = null, $base_api_path = 'rest' ) {
```

Comment:
This looks to be a copy of the method with the same [name in the `Client` class](https://github.com/Automattic/jetpack/blob/master/packages/connection/src/class-client.php#L365). Could we use that instead?